[RFC] stm32: Move prescaler from pwm binding to timers binding

ABOSTM opened the pull request:
stm32: Move prescaler from pwm binding to timers binding

Prescaler is misplaced in pwm binding. It should be in timers binding instead.
For example, TIM6/TIM7 doesn't have PWM capability, but have a prescaler.
Also prescaler is common to all PWM channels of the same timer.
So prescaler should not be in pwm binding.
This change also prepares the introduction of timer based counter
(which requires prescaler at timer level, as it is the same than for pwm)
For compatibility reason temporarily keep pwm binding to avoid
breaking boards out of tree.

ABOSTM committed:
dts: bindings: stm32: move "st,prescaler" to timers instead of pwm

Prescaler was misplaced in pwm binding, instead of timers binding.
For example, TIM6/TIM7 doesn't have PWM capability,
but have a prescaler.
Also prescaler is common to all PWM channels of the same timer.
This change also prepares the introduction of timer based counter
(which requires prescaler at timer level)
For compatibility reason temporarily keep pwm binding to avoid
breaking boards out of tree.
Block st,prescaler property in lptim binding
as lptim doesn't use this property for now.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM committed:
dts: arm: st: stm32: move "st,prescaler" to timers instead of pwm

Prescaler was misplaced in pwm binding, instead of timers binding.
For example, TIM6/TIM7 doesn't have PWM capability,
but have a prescaler.
This change also prepares the introduction of timer based counter
(which requires prescaler at timer level)

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM committed:
drivers: pwm: stm32: move "st,prescaler" to timers instead of pwm

Prescaler was misplaced in pwm binding, instead of timers binding.
For example, TIM6/TIM7 doesn't have PWM capability,
but have a prescaler.
This change also prepares the introduction of timer based counter
(which requires prescaler at timer level)
For compatibility reason temporarily use pwm prescaler if it exists,
otehrwise use timers prescaler.
And thus avoid to avoid breaking boards out of tree.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM committed:
boards: arm: stm32: move "st,prescaler" to timers instead of pwm

Prescaler was misplaced in pwm binding, instead of timers binding.
For example, TIM6/TIM7 doesn't have PWM capability,
but have a prescaler.
This change also prepares the introduction of timer based counter
(which requires prescaler at timer level)

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM committed:
tests: drivers: move "st,prescaler" to timers instead of pwm

Prescaler was misplaced in pwm binding, instead of timers binding.
For example, TIM6/TIM7 doesn't have PWM capability,
but have a prescaler.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

martinjaeger reviewed (changes requested):
This change makes a lot of sense. Only got a minor comment.

There is another slightly related flaw in the PWM driver API: The period can be defined per PWM channel in Devicetree using a phandle (see also `st,stm32-pwm.yaml`). However, it will affect all PWM channels of that timer, which might not be obvious for the user. But that would be something for a separate issue/PR.

martinjaeger reviewed (approved):
Fine with me as soon as the tiny style issue in CI is fixed.

henrikbrixandersen reviewed (approved):
This makes sens.e I have reviewed the driver change, but I have not gone through all the changed DTS files.

FRASTM reviewed (commented):
I still wonder the benefit of setting a `st,prescaler = <0>;`  for all the stm32 dtsi

erwango commented on dts/bindings/pwm/st,stm32-pwm.yaml:
```diff
@@ -10,8 +10,12 @@ properties:
+      description: |
+        Clock prescaler at the input of the timer
+        Note: This property is obsolete and was replaced by "st,prescaler" at
+        timers level. It is kept temporarily for compatibility reasons.
```
Lest' be more specific about "timer level".
It can be replaced by "binding of parent timer node" or something similar

erwango commented on dts/bindings/timer/st,stm32-timers.yaml:
```diff
@@ -19,3 +19,11 @@ properties:
+      description: |
+        Clock prescaler at the input of the timer
+        Could be in range [0 .. 0xFFFF] for STM32 General Purpose Timers (CLK/(prescaler+1) )
+        Or [0..0x3] for STM32 LPTimers  (CLK/(2^prescaler))
```
Why mentioning LPTimer if property is not available for those ?

erwango commented on dts/arm/st/l0/stm32l071.dtsi:
```diff
@@ -80,16 +80,15 @@
-				st,prescaler = <0>;
 				label = "PWM_6";
 				#pwm-cells = <3>;
-			};
```
Don't remove this "};"

erwango commented on dts/arm/st/l0/stm32l071.dtsi:
```diff
@@ -98,16 +97,15 @@
-				st,prescaler = <0>;
 				label = "PWM_7";
 				#pwm-cells = <3>;
-			};
```
Don't remove this "};"

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -356,9 +356,8 @@ replaced by 'st,prescaler' property in parent node, aka timers"
-		/* For compatibility reason, use pwm st_prescaler property if  \
-		 * exist otherwise use parent (timers) property                \
-		 */                                                            \
+		/* For compatibility reason, use pwm st_prescaler property  */ \
```
Move this change to initial driver change commit

mbolivar-nordic reviewed (commented):
no objections

ABOSTM committed:
dts: arm: stm32: TIM6 and TIM7 doesn't support PWM capability

TIM6 and TIM7 doesn't have comparator, nor pin output.
Thus they can't support PWM.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented:
> Lest' be more specific about "timer level".
> It can be replaced by "binding of parent timer node" or something similar

Done

> Move this change to initial driver change commit

Done


> Why mentioning LPTimer if property is not available for those ?

Initially it was common to LPTIM. Now this property is blocked for LPTIM.
So you are right it should be removed. Done

> Don't remove this "};"

In fact it is removed but now, it is done in the right commit.

ABOSTM commented:
Rebase to fix buildkite issue (linked to uart).
Temporary increase buildkite timeout.

erwango commented:
@galak, @cfriedt, @MaureenHelm This one is ready to merge once the temp patch is removed.

ABOSTM commented:
CI is passed. Removing Buildkite timeout patch.

MaureenHelm commented:
Please remove the `DNM` label if this is indeed ready to merge.

erwango commented:
@MaureenHelm, now ready
